test/system/255-auto-update.bats

```diff
@@ -0,0 +1,308 @@
+function generate_service() {
+    target_img_basename=$1
+    autoupdate=$2
+
+    # Generate a random container name and echo it to output. By this way
+    # we can pass the container name to the tests. Please keep this is the
+    # first echo from this function to make it possible that tests can get
+    # the container name and use it in the following steps.
+    cname=c_$(random_string)
+    echo $cname
+    target_img="quay.io/libpod/$target_img_basename:latest"
+    run_podman tag $IMAGE $target_img
+    if [[ -n "$autoupdate" ]]; then
+        label="--label io.containers.autoupdate=$autoupdate"
+    fi
+    run_podman run -d --name $cname $label $target_img top -d 120
+
+    run_podman generate systemd --new $cname
+    echo "$output" > "$UNIT_DIR/container-$cname.service"
+    echo "container-$cname" >> $SNAME_FILE
+    run_podman rm -f $cname
+
+    systemctl daemon-reload
+
+    run systemctl start container-$cname
+    if [ $status -ne 0 ]; then
+        die "Error starting systemd unit $SERVICE_NAME, output: $output"
+    fi
+
+    run systemctl status container-$cname
+    if [ $status -ne 0 ]; then
+        die "Non-zero status of systemd unit $SERVICE_NAME, output: $output"
+    fi
+}
+
+function _wait_service_ready() {
+    local sname=$1
+
+    local timeout=5
+    while [[ $timeout -gt 0 ]]; do
+        run systemctl is-active $sname
+        if [[ $output == "active" ]]; then
+            return
+        fi
+        sleep 1
+        let timeout=$timeout-1
+    done
+
+    # Print serivce status as debug information before failed the case
+    run systemctl status $sname
+    echo $output
+    die "Timed out waiting for $sname to start"
+}
+
+function _confirm_update() {
+    local sname=$1
+
+    local timeout=5
+    last_log=""
+    while [[ $timeout -gt 0 ]]; do
+        if [[ "$output" == "$last_log" ]]; then
+            return
+        fi
+        last_log=$output
+        sleep 1
+        let timeout=$timeout-1
+    done
+
+    die "Timed out waiting for $sname to update"
+}
+
+# This test can fail in dev. environment because of SELinux.
+# quick fix: chcon -t container_runtime_exec_t ./bin/podman
+@test "podman auto-update - label io.containers.autoupdate=image" {
+    run_podman images
+    run generate_service alpine image
+    cname=${lines[0]}
+    echo $output
+    run_podman inspect --format "{{.Image}}" $cname
+    ori_image=$output
+
+    _wait_service_ready container-$cname.service
+    run_podman ps -a
+    run_podman auto-update
+    is "$output" "Trying to pull.*" "Image is updated."
+    run_podman ps -a
+    _confirm_update container-$cname.service
+    run_podman inspect --format "{{.Image}}" $cname
+    [[ "$output" != "$ori_image" ]]
+}
+
+@test "podman auto-update - label io.containers.autoupdate=disabled" {
+    run generate_service alpine disabled
+    cname=${lines[0]}
+    echo $output
+    run_podman inspect --format "{{.Image}}" $cname
+    ori_image=$output
```

Comment:
There's a lot of duplication of this five-line stanza. Did you know that bash variables aren't scoped? That is, you can define a variable in a function, and the caller will be able to see it. (You can't use `run` though). So if you do the `podman inspect + ori_image=` inside `generate_service`, you can make all these five-line repeats become just one line.

Reply:
Thanks a lot for bring this out. Then I think I can get ride of "echo $cname" inside the function too. Will update the code.